File: solutions/fsharp/book-store/BookStore.fs
module BookStore

let total books =
    // Initialize an array to hold the counts of each book series
    let series = Array.zeroCreate 5
    
    // Count the occurrences of each book series
    books
    |> Seq.countBy id
    |> Seq.iter (fun (i, count) -> series[i - 1] <- count)
    Array.sortInPlace series
    
    // Calculate the differences between each count and the previous one    
    let calculateDifference i x = 
        if i = 0 then x else x - series[i - 1]
    let groups = Array.mapi calculateDifference series
    
    // Combine groups to maximize the discount
    let combined = min groups[2] groups[0]
    groups[0] <- groups[0] - combined
    groups[2] <- groups[2] - combined
    groups[1] <- groups[1] + combined * 2
    
    // Define the prices for each group size
    let groupPrice = [
        5 * 8 * 75
        4 * 8 * 80
        3 * 8 * 90
        2 * 8 * 95
        1 * 8 * 100
    ]
    let calculatePrice i x = 
        decimal (x * groupPrice[i]) / 100m
        
    // Calculate the total price
    groups |> Array.mapi calculatePrice |> Array.sum
    

File: solutions/scala/book-store/src/main/scala/BookStore.scala
object BookStore:
  private val BookPrice = Map(1 -> 800, 2 -> 760, 3 -> 720, 4 -> 640, 5 -> 600)
  private val SeriesCount = 5

  def total(books: List[Int]): Int =
    // Count the occurrences of each book series
    val seriesCounts = (1 to SeriesCount).map(i => books.count(_ == i)).sorted.toArray

    // Calculate the differences between each count and the previous one, reversed
    val groupCounts = (0 until SeriesCount).map { i =>
      if i == 0 then seriesCounts(0) else seriesCounts(i) - seriesCounts(i - 1)
    }.reverse.toArray

    // Rearrange books in groups to get maximum discount
    val combined = math.min(groupCounts(2), groupCounts(4))
    val adjustedGroups = groupCounts
      .updated(4, groupCounts(4) - combined)
      .updated(2, groupCounts(2) - combined)
      .updated(3, groupCounts(3) + 2 * combined)

    // Calculate the total price with maximum discount
    (1 to SeriesCount).map { i => i * BookPrice(i) * adjustedGroups(i - 1) }.sum
